### PR TITLE
Add e2e reduction fusion tests

### PR DESCRIPTION
**Pull request by okkwon**
1. reduction + broadcast + elementwise
2. softmax
3. layernom

Note that the tests have static shapes.

### CODE REVIEW

**Comment by okkwon** [comment]
@pzread, do you know a good place to add dynamic cases? How can I pass large inputs like 12x128x128 for <?x?x?> or <12x?x?>? For the static cases, I used constants, but for dynamic cases, it looks like that I have to pass the inputs through `iree-run-mlir`, but I am not clear how to pass a large input.

**Comment by okkwon** [comment]
Issue: https://github.com/iree-org/iree/issues/9938

**Comment by pzread** [comment]
> @pzread, do you know a good place to add dynamic cases? How can I pass large inputs like 12x128x128 for or <12x?x?>? For the static cases, I used constants, but for dynamic cases, it looks like that I have to pass the inputs through `iree-run-mlir`, but I am not clear how to pass a large input.

Probably not the most knowledgeable person to answer this question, but looks like there are some useful patterns under https://github.com/iree-org/iree/tree/main/tests/e2e/regression

[One](https://github.com/iree-org/iree/blob/main/tests/e2e/regression/dynamic_add.mlir) is using `flow.tensor.constant`, [the other](https://github.com/iree-org/iree/blob/main/tests/e2e/regression/dynamic_linalg_matmul_on_tensors.mlir) is using `lit` with `iree-run-mlir`

**Comment by okkwon** [comment]
Thanks for the info.

Using lit seems limited because I need to pass the whole tensor values. (I am not aware of a way to pass a scalar constant for a large tensor through lit.

I will try the constant method. I didn't know that I can get a dynamic shaped tensor from a constant value, which is quite interesting. I wonder if the same thing can be done with the unfoldable constant.

@benvanik do you have an idea?

File: tests/e2e/patterns/BUILD [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    [
+        "layernorm.mlir",
+        "reduction_broadcast_elementwise.mlir",
+        "softmax.mlir",
+    ],
+    include = ["*.mlir"],
+)
+
+iree_check_single_backend_test_suite(
+    name = "check_dylib-llvm-aot_dylib",
```
Comment (ScottTodd) [inline tests/e2e/patterns/BUILD:29]: Please use the new name here - `check_llvm-cpu_local-task` (see https://github.com/iree-org/iree/pull/9934 for how other tests were updated)

**Comment by okkwon** [comment]
I will move the tests to the `regression` folder.

**Comment by ScottTodd** [comment]
> How can I pass large inputs like 12x128x128 for or <12x?x?>? For the static cases, I used constants, but for dynamic cases, it looks like that I have to pass the inputs through iree-run-mlir, but I am not clear how to pass a large input.

> Using lit seems limited because I need to pass the whole tensor values. (I am not aware of a way to pass a scalar constant for a large tensor through lit.
> 
> I will try the constant method. I didn't know that I can get a dynamic shaped tensor from a constant value, which is quite interesting. I wonder if the same thing can be done with the unfoldable constant.

Some of your options are:
* `flow.tensor.constant` (explicit constant, can return a dynamic shape), e.g. https://github.com/iree-org/iree/blob/6656273a5e0ed09f70d8d697962031b6715ba9ee/tests/e2e/regression/dynamic_abs.mlir#L1-L8
* `flow.tensor.splat` (creates a tensor filled with a single value), e.g. https://github.com/iree-org/iree/blob/6656273a5e0ed09f70d8d697962031b6715ba9ee/compiler/src/iree/compiler/Dialect/Stream/Conversion/FlowToStream/test/tensor_ops.mlir#L61-L69

File: tests/e2e/patterns/BUILD (hunk 1 above)
Comment (okkwon) [reply tests/e2e/patterns/BUILD]: Thanks. I moved the tests into `regression` and follow the name pattern there.

**Comment by okkwon** [comment]
Thanks @ScottTodd. I'd like to make the test self-contained, so the first option sounds good to me.

**Review by ThomasRaoux — APPROVED** [review]
Thanks for adding those!

**Review by ScottTodd — APPROVED** [review]
I scanned through the test results to make sure these tests won't be adding a unreasonable amount of time to our CI pipelines. Looks good:
```
iree/tests/e2e/regression                                        =  14.75 sec*proc (52 tests)
```
(If that were > 1 minute we'd want to take a closer look at the dimensions used and consider disabling the tests on some backends / configurations)

**Comment by ScottTodd** [comment]
> * reduction + broadcast + elementwise
> * softmax
> * layernom

Also just curious: were there bugs in any of these that these new tests would have caught? We'd like for the "regression" folder to contain that sort of test. Or, to put it another way: what specific value do these tests provide over unit tests for the underlying operations?

**Comment by okkwon** [comment]
> Also just curious: were there bugs in any of these that these new tests would have caught? We'd like for the "regression" folder to contain that sort of test. Or, to put it another way: what specific value do these tests provide over unit tests for the underlying operations?

These are one of the important fusion areas, and having the small e2e tests are useful to keep the compilation sanity and to track the performance.

**Comment by benvanik** [comment]
Regression tests are generally the dumping ground for when a reported bug can't be understood or broken down any further and not intended for new functional tests. For anything that does end up in regression the intent is that over time as proper tests are added in the appropriate places in the compiler it can be pruned/deleted; so adding new things there should be seen only as temporary test coverage and not functional coverage. These particular tests may make sense to have under an e2e/tosa_layers/ or something path and from the source dialect tosa ops - maintaining these as is is going to be difficult and they're likely to get deleted when they cannot be updated (the tosa->linalg lowerings change and they're no longer representative of any input we actually see, etc).

We can make the intent clearer by making all tests under regression/ have an issue number in them. Having tosa_layers/ to match tosa_ops/ seems like the home for these new ones. I'll do some shuffling.

**Comment by okkwon** [comment]
Thanks for the comment. One thing that didn't fit very well was the dynamic shape with TOSA. I was not very clear about the dynamic shape support from TOSA, and when I tried a case, the compiler reports an unrelated error. I'd much prefer writing a test in TOSA since it is easier to write and more succint.

**Comment by benvanik** [comment]
Ah good question - you can use `flow.tensor.constant` to get dynamically shaped constants that won't fold:
```mlir
// encoded as 2x3xf32 but treated as ?x?xf32 - could partially shape as well (2x?xf32)
%constant = flow.tensor.constant dense<[[-1.0, 2.0, -3.0], [4.0, -5.0, 6.0]]> : tensor<2x3xf32> -> tensor<?x?xf32>
```

**Comment by okkwon** [comment]
> Ah good question - you can use `flow.tensor.constant` to get dynamically shaped constants that won't fold:

Yes, I used that in the dynamic cases. We can move the test to tosa_layers to use the static tests in TOSA while keeping the dynamic tests in Linalg.

**Comment by okkwon** [comment]
Just a side note: I tried mhlo to write the tests, but TOSA was much easier to write. TOSA seems to be a good high level IR for manual IR typing.

**Comment by benvanik** [comment]
I've found the same :)
Having more of these layer-wise tests will be handy - thanks for starting this!

**Pull request merged**